vm/src/stdlib/array.rs

```diff
@@ -435,12 +534,36 @@ impl PyArray {
 
     #[pymethod(magic)]
     fn setitem(
-        &self,
+        zelf: PyRef<Self>,
         needle: Either<isize, PySliceRef>,
         obj: PyObjectRef,
         vm: &VirtualMachine,
     ) -> PyResult<()> {
-        self.borrow_value_mut().setitem(needle, obj, vm)
+        match needle {
+            Either::A(i) => zelf.borrow_value_mut().setitem_by_idx(i, obj, vm),
+            Either::B(slice) => {
+                let cloned;
+                let guard;
+                let items = if zelf.is(&obj) {
+                    cloned = zelf.borrow_value().clone();
```

Comment:
is `derive(Clone)` required by this one?

Reply:
Yes, it clone out value from zelf, we need to modify zelf later so we need a copy, a ArrayContentType enum with a Vec. Is any reason we don't use derive(Clone)?

Reply:
Can we copy it *after* slicing the original? It seems it is wasting copy cost a little bit

Reply:
We must use all of the items that supplied on the right side of operator, otherwise it is a err. If step is 1, we do replace and insert(splice). If step is not 1, the length of slice must equal to length of items. So here should have no waste unless the err happend.

Reply:
Oh, that's right.